[GEN][ZH] Unify code of GameAudio and MilesAudioManager

Mauller opened the pull request:
**Squash and merge**

This PR makes the changes required to synchronize the audio implementations between Generals and Zero hour.

Most changes are backports to Generals, while some small tweaks are made to zero hour to match generals.

- Merge After: #793

Mauller commented on Generals/Code/GameEngineDevice/Include/MilesAudioDevice/MilesAudioManager.h:
```diff
@@ -53,12 +53,12 @@ enum PlayingWhich CPP_11(: Int)
-	union
-	{
+//	union
+//	{
 		HSAMPLE m_sample;
 		H3DSAMPLE m_3DSample;
 		HSTREAM m_stream;
-	};
+//	};
```
I think the generals way of handling this might have been better on memory.

Playing audio will only ever be a single type.

xezon replied:
Yes it looks like it should be a union. The difference is that in Zero Hour these are 0 initialized. But in Generals they are probably also zero initialized because of the Game Memory.

Mauller replied:
I will revert the zero hour way of doing it an Null initialise one of the members in both.

That should make sure the members are initialised then.

Mauller commented on GeneralsMD/Code/GameEngineDevice/Include/MilesAudioDevice/MilesAudioManager.h:
```diff
@@ -170,6 +170,9 @@ class MilesAudioManager : public AudioManager
 
 		virtual void killAudioEventImmediately( AudioHandle audioEvent );
 
+		virtual void stopAllAmbientsBy( Object *objID );
+		virtual void stopAllAmbientsBy( Drawable *drawID );
+
```
Although the functionality for these is empty at the moment, it appears that the idea was to use them to stop audio ambient audio in specific circumstances.

Ambient audio seems to be constantly playing sound effects that occur in the background. These have some code within generals that would have paused them on the menu.

The other situation could be that they get paused when out of range etc. But i think audio culling is handled elsewhere.

xezon replied:
Perhaps this was just legacy code that was meant to be removed? There are these 2 functions:
```
Drawable::startAmbientSound
Drawable::stopAmbientSound
```
I suspect these are all we need to start and stop ambient audio on drawables?

## Generals

In Generals there are these 2 calls to stopAllAmbientsBy:
```cpp
void GameLogic::setGamePaused( Bool paused, Bool pauseMusic )
	...
	if(paused)
	{
	...
	}
	else
	{
	...
		while( drawable )
		{
			drawable->startAmbientSound();
			TheAudio->stopAllAmbientsBy( drawable );
			drawable = drawable->getNextDrawable(); // <--- This call makes no sense
		}
	}
...
```
```cpp
void FXListDie::onDie( const DamageInfo *damageInfo )
{
	...
	if (d->m_defaultDeathFX)
	{
		// if the object has any ambient sound(s), kill 'em now.
		TheAudio->stopAllAmbientsBy(getObject());
		
		if (d->m_orientToObject)
		{
		...
```

## Zero Hour

```
void FXListDie::onDie( const DamageInfo *damageInfo )
{
	...
	if (d->m_defaultDeathFX)
	{
		if (d->m_orientToObject)
		{
		...
}
```

I suggest to remove these virtual functions from Generals because they have no use in both titles.

Mauller replied:
Yeah, if we don't think they are useful In the end I will remove them from both variants.

Will then make a followup PR to correct any code that used them.

Will also check the situation with the FXList variant wanting to kill ambient sounds on an object, I don't think Generals or zero hour puts ambient sounds on objects and only drawables.

Mauller replied:
Yeah looks like only drawables have ambient sounds on them.

So the object version of the function would have needed to check if the object had a drawable, then it would have to disable the sound on the drawable if it had one.

Mauller commented:
Updated and addressed the discussed points above, will push a small followup PR to remove the stragling bits of code associated with the removed interface functions from GameAudio.

Mauller commented:
Had to make a small push again to include the removal of calls to ```stopAllAmbientSoundsBy()``` since the build pipeline would fail otherwise for the generals build.

xezon commented:
Is this change meant to be merged with rebase or not?

Commit number 3 of this change appears to error:
```
Error: D:\a\GeneralsGameCode\GeneralsGameCode\Generals\Code\GameEngine\Source\GameLogic\Object\Die\FXListDie.cpp(69) : error C2039: 'stopAllAmbientsBy' : is not a member of 'AudioManager'
```

Mauller commented:
> Is this change meant to be merged with rebase or not?
> 
> Commit number 3 of this change appears to error:
> 
> ```
> Error: D:\a\GeneralsGameCode\GeneralsGameCode\Generals\Code\GameEngine\Source\GameLogic\Object\Die\FXListDie.cpp(69) : error C2039: 'stopAllAmbientsBy' : is not a member of 'AudioManager'
> ```

The build of this PR should be fixed with the last commit i made to add removing those calls to ```stopAllAmbientsBy```

since MilesAudioManager inherits from GameAudio, those two commits are kind of interdependent, so i thought it would be better to squash and merge this. But merging by rebase should be okay.

xezon commented:
If commit number 3 fails to compile on its own, then it cannot be merged with rebase.

Mauller commented:
> If commit number 3 fails to compile on its own, then it cannot be merged with rebase.

Reordered the commits to break the problematic dependency.

sorcerer86pt reviewed (approved):
LGTM

xezon commented on Generals/Code/GameEngine/Include/Common/GameAudio.h:
```diff
@@ -152,10 +152,6 @@ class AudioManager : public SubsystemInterface
 
-		// device dependent stops.
-		virtual void stopAllAmbientsBy( Object* obj ) = 0;
-		virtual void stopAllAmbientsBy( Drawable* draw ) = 0;
```
There are now 3 commits that remove pieces of `stopAllAmbientsBy` in different places. How about put the removal for `stopAllAmbientsBy` in one commit? I suggest create one new pull request where this `stopAllAmbientsBy` stuff is removed first, then we see next steps.

Mauller replied:
Seperated that out and into #793

Mauller commented:
This will fail at the moment till https://github.com/TheSuperHackers/GeneralsGameCode/pull/793 is merged before it.

xezon commented:
Can we merge this with Squash?

Mauller commented:
> Can we merge this with Squash?

Yeah should be fine, can encapsulate the changes in one.

xezon reviewed (approved):
Ok. Adjusted the title of this change.

Mauller commented:
Rebased with main so good to go.
